docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,81 @@
+
+Pitfalls of current solutions:
+* It’s tedious to manually visit each leaf and we are lazy.
+* Visiting leaf instances (be it ThanosRuler or dedicated Prometheus endpoints) have to be implemented manually.
+
+## Goals
+
+* Present other Prometheus/Thanos Resources in a global view.
+    * This also means an up-to-date view (e.g statuses)
+* It can be tied to Thanos.
+* Simple to run if you already run Thanos.
+    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
+    * Potentially use the same connections, discovered targets by Querier.
+
+## Verification
+
+* Unit tests which would fire up a dummy RulesAPI and check different scenarios.
+* Ad-hoc testing.
+
+## Proposal
+
+We propose to add following proto service and propagate all using existing components that has rules, so:
+* Sidecar
+* Querier (for federation)
+* Ruler
+
+The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alertings rules.
+
+```proto
+service Rules {
+  /// Rules has info for all rules.
+  rpc Rules(RulesRequest) returns (stream RulesResponse);
+}
+```
+
+Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly Thanos rule constructs the response based in its local state.
+
+Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:
+
+* Deduplication happens first the rule group level. The identifier is the group name.
+* Then, per group name deduplication happens on the rule level, where:
+
+1. the rule type (recording rule vs. alerting rule)
+2. the rule name
+3. the rule label names
+
+are being used as the deduplication identifer. Disjunct entries are simply merged by adding them to the result set.
+
+Thanos query presents the result of the fan-out on a `/api/v1/rules` endpoint which is compatible with the Prometheus' rules API endpoint.
+
+## Alternatives
+
+* Cortex contains a sharded Ruler. Assigning rules to shards is done via Consul, though a gossip implementation is under development. Shards do not communicate with other shards. Rules come from a store (e.g. a Postgres database).
+
+## Work Plan
+
+* Implement a new flag `--rule` in Thanos Query which registers RulesAPI endpoints.
+* Implement RulesAPI backends in sidecar, query, rule.
+* Other StoreAPI backends implement a Noop RulesAPI backend, error'ing on RulesAPI invocations.
```

Comment:
Why do we need a Noop Rules API? Can we just not implement it?

Reply:
I don't think so, because other Thanos components do not have any means of providing Rules API.

Reply:
Actually that's a good catch @povilasv  - Why we need to implement anything for those? Just do nothing for those (:

Reply:
I missed the "not" :man_facepalming: :-D

Reply:
:smile: seems like this was solved so marking as "Resolved"

Reply:
This sentence is still wrong :/
There is no need to do `Other StoreAPI backends implement a Noop RulesAPI backend, error'ing on RulesAPI invocations.` 

can we just remove it?

Reply:
ah yes, indeed, removed.